lib/galaxy/jobs/runners/util/cli/shell/rsh.py

```diff
@@ -26,9 +26,9 @@ def __init__(self, rsh='rsh', rcp='rcp', hostname='localhost', username=None, **
     def execute(self, cmd, persist=False, timeout=60):
         # TODO: implement persistence
         if self.username is None:
-            fullcmd = '%s %s %s' % (self.rsh, self.hostname, cmd)
+            fullcmd = '%s %s "%s"' % (self.rsh, self.hostname, cmd)
         else:
-            fullcmd = '%s -l %s %s %s' % (self.rsh, self.username, self.hostname, cmd)
+            fullcmd = '%s -l %s %s "%s"' % (self.rsh, self.username, self.hostname, cmd)
```

Comment:
I don't think this works because we construct a ssh command here, that's why we need shell="true" as well in https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/jobs/runners/util/cli/shell/local.py#L47

Reply:
actually it might work, I had tried something else before.

Reply:
Yep, you can set `shell=True` only if `cmd` is a string.

Reply:
Hmm, is this what you had in mind?
```
diff --git a/lib/galaxy/jobs/runners/util/cli/shell/local.py b/lib/galaxy/jobs/runners/util/cli/shell/local.py
index a5b570abaa..7e6b65c521 100644
--- a/lib/galaxy/jobs/runners/util/cli/shell/local.py
+++ b/lib/galaxy/jobs/runners/util/cli/shell/local.py
@@ -44,7 +44,7 @@ class LocalShell(BaseShellExec):
 
     def execute(self, cmd, persist=False, timeout=DEFAULT_TIMEOUT, timeout_check_interval=DEFAULT_TIMEOUT_CHECK_INTERVAL, **kwds):
         outf = TemporaryFile()
-        p = Popen(cmd, shell=True, stdin=None, stdout=outf, stderr=PIPE)
+        p = Popen(cmd, stdin=None, stdout=outf, stderr=PIPE)
         # poll until timeout
 
         for i in range(int(timeout / timeout_check_interval)):
diff --git a/lib/galaxy/jobs/runners/util/cli/shell/rsh.py b/lib/galaxy/jobs/runners/util/cli/shell/rsh.py
index d1b3b5f74f..c3e370dfc1 100644
--- a/lib/galaxy/jobs/runners/util/cli/shell/rsh.py
+++ b/lib/galaxy/jobs/runners/util/cli/shell/rsh.py
@@ -26,9 +26,11 @@ class RemoteShell(LocalShell):
     def execute(self, cmd, persist=False, timeout=60):
         # TODO: implement persistence
         if self.username is None:
-            fullcmd = '%s %s "%s"' % (self.rsh, self.hostname, cmd)
+            fullcmd = [self.rsh, self.hostname, "'%s'" % cmd]
         else:
-            fullcmd = '%s -l %s %s "%s"' % (self.rsh, self.username, self.hostname, cmd)
+            fullcmd = [self.rsh, '-l', self.username, self.hostname, "'%s'" % cmd]
         return super(RemoteShell, self).execute(fullcmd, persist, timeout)
```

This fails with `OSError: [Errno 2] No such file or directory`

Reply:
Ah, this is because of the additional options, got it.

Reply:
Awesome, no need for this hackish fix then! 🏅 to 🦅 👁 !